nemo_aligner/utils/utils.py

```diff
@@ -122,6 +122,7 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
     if remove_meta_info:
         checkpoint_cfg.pop("target", None)
         checkpoint_cfg.pop("nemo_version", None)
+        checkpoint_cfg.data.pop("data_prefix", None)
```

Comment:
> when they start with a pretrained checkpoint, the data_prefix turned into a list instead of dictionary

Oh I see, yeah having an option that can either be a list or a dict isn't great for overriding...

> I feel like this change is too hacky, instead we should modify all our conf such as [gpt_ppo_actor](https://github.com/NVIDIA/NeMo-Aligner/blob/main/examples/nlp/gpt/conf/gpt_ppo_actor.yaml#L168) to just define it's own data field so it overwrites the one inside the model

I'm also not a fan of having it under the `if remove_meta_info` block since this is not really "meta" info.

However, it's not just a matter of modifying the configs. The [training_rm](https://github.com/NVIDIA/NeMo-Aligner/blob/20afb25da273ccdda3341df95d0a71a9bc888c9d/examples/nlp/gpt/conf/training_rm.yaml#L95) already define its own `data` field, but the way [OmegaConf.merge()](https://github.com/NVIDIA/NeMo-Aligner/blob/6baf3b1d0613896069a63e70773a1eb194c4c89a/nemo_aligner/utils/utils.py#L126) works, it's going to recurse into `data_prefix` rather than discard the full original `data` config from the model, which will crash if the former is a dict while the latter is a list (note also that if the latter was a dict, it wouldn't crash but could do weird things as the existing keys would be kept).

I'm afraid there's no great solution to this problem because we need this merge behavior (that's what allows us to avoid copying the full model config as we used to need in the past). But there are config nodes that we'd rather overwrite entirely rather than do a recursive merge: `data_prefix` is one, but there may be others we're not seeing yet, or that may appear later (I actually wonder if we should overwrite the whole `data` section for RM training for instance -- otherwise we carry around a lot of irrelevant options coming from the base model).

Here's a potential idea to address this (better suggestions welcome -- but I'd rather avoid hardcoding this into the code, as (1) it's a bit obscure and (2) we may encounter new use cases that require being handled differently):
* Add a new config section `overwrite_base_config` (or `ignore_model_config`, or any better name) that lists the config fields that should be overwritten rather than merged, e.g.
```yaml
overwrite_base_config:
  model:
    data:
      data_prefix: True
```
* For these fields, pop them from the base config before the merge when they're also present in the config being merged
* The motivation for this spec format is that it's pretty flexible, and if we end up using similar fields in multiple configs we could move it to a common config imported in the `defaults` list (also to avoid cluttering the configs)

Again, happy to go with a better solution if someone has one in mind!

Reply:
> However, it's not just a matter of modifying the configs. The [training_rm](https://github.com/NVIDIA/NeMo-Aligner/blob/20afb25da273ccdda3341df95d0a71a9bc888c9d/examples/nlp/gpt/conf/training_rm.yaml#L95) already define its own data field, but the way [OmegaConf.merge()](https://github.com/NVIDIA/NeMo-Aligner/blob/6baf3b1d0613896069a63e70773a1eb194c4c89a/nemo_aligner/utils/utils.py#L126) works, it's going to recurse into data_prefix rather than discard the full original data config from the model, which will crash if the former is a dict while the latter is a list (note also that if the latter was a dict, it wouldn't crash but could do weird things as the existing keys would be kept).

Yep, that's exactly the issue, and from what I could tell, indeed there's no great solution based on directly overriding the configs. Adding a `overwrite_base_config` section sounds good to me. If there isn't any other suggestion I can go ahead and implement it.

Reply:
ah, that is tricky. The above proposal sounds good, but I wonder why we even have the data attribute bound to the model in aligner. This is a thing in NeMo.

What if we just move the `data` field out of the model attribute and only access that? That would make it cleaner but i think users will be confused if they come from nemo

Reply:
> What if we just move the `data` field out of the model attribute and only access that?

That's indeed a potential option, but it also has drawbacks:
1. Could be confusing as you mentioned, especially when we dump our configs that will end up having two unrelated `data` nodes (one under `model` and one outside)
2. It would require changing the model code that currently accesses `self.cfg.data`. There isn't a lot but I'm not sure how easily we can get rid of it (I haven't looked at it very closely).

I'm not against it though (maybe point 2 can simply be addressed with config interpolations), as it could be a simple fix. That being said, I was thinking it may be useful to have a more generic mechanism as this could prove useful in some other situations as well. For instance it probably doesn't make much sense to merge the `optim` section either, even if it hasn't caused problems yet.

Reply:
good point, i think having this explicitly also reduces the chance for confusion.

OK to go with adding a `overwrite_base_config` and then doing as @odelalleau  mentioned above

Reply:
@gleibovich-nvidia let me know if you'd like me to implement this feature

Reply:
Thanks @odelalleau and @gshennvm. 
As for implementation, if you have the time, sure go ahead. Otherwise I can get to it sometime during next week.

Reply:
I should be able to work on it on Monday.
Just to be sure we don't overlap work, I created a related issue #53 => whoever starts working on it first can grab it :)